File: problems/DSA1/sorting/2.noble-integer.js
/**
 * Run this problem:
 * node problems/DSA1/sorting/2.noble-integer.js
 */

/**
 * Problem: Noble Integer
 * Difficulty: Easy
 * Category: Sorting
 * 
 * Description:
 * Given an integer array A, find if an integer p exists in the array such that 
 * the number of integers greater than p in the array equals p.
 * 
 * Constraints:
 * 1 <= |A| <= 2*10^5
 * -10^8 <= A[i] <= 10^8
 * 
 * Examples:
 * Input: [3, 2, 1, 3] -> Output: 1 (p=2 exists)
 * Input: [1, 1, 3, 3] -> Output: -1
 */

/**
 * Approach: Sorting
 * 
 * Algorithm:
 * 1. Sort the array in ascending order.
 * 2. Traverse the array and check:
 *    - If A[i] == A[i+1], skip (count of elements greater than A[i] is same as A[i+1]).
 *    - The number of elements greater than A[i] is (total_length - 1 - i).
 *    - If A[i] == (total_length - 1 - i), then A[i] is noble.
 * 3. Special case for the last element:
 *    - If A[n-1] == 0, it is noble (0 elements greater than it).
 * 
 * Time Complexity: O(N log N) for sorting.
 * Space Complexity: O(1) in-place sort.
 */
function solve(A) {
    if (!A || A.length === 0) return -1;

    // Create a copy to avoid mutating input passed by reference
    const arr = [...A].sort((a, b) => a - b); // Time: O(N log N) | Space: O(N)
    const n = arr.length;

    for (let i = 0; i < n - 1; i++) {
        // Skip duplicate elements
        if (arr[i] === arr[i + 1]) continue;

        // Count of numbers greater than arr[i]
        const countGreater = n - 1 - i;
        if (arr[i] === countGreater) {
            return 1;
        }
    }

    // Check last element
    if (arr[n - 1] === 0) {
        return 1;
    }

    return -1;
}

// ============================================
// RUN TESTS
// ============================================
const { runTests } = require('../../../utils/test-runner');
const { testCases, validator } = require('../../../utils/DSA1/sorting/2.noble-integer');

if (require.main === module) {
    // 1. Run Automated Test Runner
    runTests({
        problemName: "Noble Integer",
        solution: solve,
        testCases: testCases,
        validator: validator,
        currentComplexity: { time: 'O(n log n)', space: 'O(1)' },
        optimalComplexity: { time: 'O(n log n)', space: 'O(1)' }
    });

    // 2. Dummy Data Execution Block (Visual Check)
    console.log("\n--- Standalone Execution ---");
    const dummyInput = [3, 2, 1, 3];
    console.log("Input:", dummyInput);
    console.log("Output:", solve(dummyInput));
}

module.exports = { solve };

/*
 * Educational Summary:
 * 1. Array.prototype.sort(): Built-in JavaScript sort (TimSort). Time Complexity is O(N log N).
 * 2. Spread Operator [...A]: Used to clone the array. Time Complexity is O(N). Space is O(N).
 * 3. Noble Integer Condition: Exploits the property of sorted arrays where n-1-i represents elements strictly to the right.
 */

File: problems/DSA1/sorting/2.NobleInteger.java
/**
 * Run:
 * & "C:\Program Files\Java\jdk-25.0.2\bin\javac.exe" problems/DSA1/sorting/2.NobleInteger.java
 * & "C:\Program Files\Java\jdk-25.0.2\bin\java.exe" problems.DSA1.sorting.NobleInteger
 * 
 * NOTE: Java is a compiled language. 
 * - When you modify this .java file, you MUST re-run 'javac' to update the .class file (bytecode).
 * - Running the code executes the .class file, not the .java file directly.
 */
package problems.DSA1.sorting;

import java.util.Arrays;

/**
 * Problem: Noble Integer
 * Difficulty: Easy
 * Category: Sorting
 * 
 * Description:
 * Given an integer array A, find if an integer p exists in the array such that 
 * the number of integers greater than p in the array equals p.
 */
class NobleInteger {

    /**
     * Finds if a noble integer exists in the array.
     * 
     * Time Complexity: O(N log N)
     * Space Complexity: O(1)
     */
    public static int solve(int[] A) {
        if (A == null || A.length == 0) return -1;

        // Sort ascending
        Arrays.sort(A); // Time: O(N log N) | Space: O(log N)
        int n = A.length;

        for (int i = 0; i < n - 1; i++) {
            // Skip duplicates
            if (A[i] == A[i + 1]) continue;

            int countGreater = n - 1 - i;
            if (A[i] == countGreater) {
                return 1;
            }
        }

        // Special case: Last element
        if (A[n - 1] == 0) {
            return 1;
        }

        return -1;
    }

    public static void main(String[] args) {
        // Dummy Data & Verification
        int[] input1 = {3, 2, 1, 3};
        int expected1 = 1;
        int result1 = solve(input1);
        System.out.println("Test 1 (Example 1): " + (result1 == expected1 ? "PASS" : "FAIL - Got " + result1));

        int[] input2 = {1, 1, 3, 3};
        int expected2 = -1;
        int result2 = solve(input2);
        System.out.println("Test 2 (Example 2): " + (result2 == expected2 ? "PASS" : "FAIL - Got " + result2));

        int[] input3 = {0, 0, 0};
        int expected3 = 1;
        int result3 = solve(input3);
        System.out.println("Test 3 (Multiple Zeros): " + (result3 == expected3 ? "PASS" : "FAIL - Got " + result3));
    }
}

/*
 * Educational Summary:
 * 1. Arrays.sort(int[]): Standard Java sorting. Time Complexity O(N log N).
 * 2. Count of Greater Elements: In a sorted array, the count of elements to the right of index i
 *    is exactly n - 1 - i.
 * 3. Handling Duplicates: By skipping if A[i] == A[i+1], we ensure the condition is only checked 
 *    at the last occurrence of a specific value.
 */
